### ISSUE TITLE
SWR AVX512 support

### ISSUE THREAD

**Issue by oscarbg** [pal1000/mesa-dist-win#2]
Phoronix found some difficulties to build AVX512 support so just asking? Perhaps fixed in 17.3 master branch?

**Comment by pal1000** [comment]
SWR AVX512 is only available in master branch.

I only publish stable releases made with Scons and Visual Studio, but nothing stops you from trying to build from a development branch / master.
Just clone mesa source code with git instead of downloading a stable release:
git clone --depth=100 --branch=master git://anongit.freedesktop.org/mesa/mesa mesa
All the other steps of [build script documentation](https://github.com/pal1000/mesa-dist-win/blob/master/buildscript/readme.md) are unchanged. 
Unlike Linux autotools where you can build with LLVM 6.0 SVN, Scons builds immediately break with new major LLVM version and they remain broken long after LLVM release. See Mesa bugs [100201](https://bugs.freedesktop.org/show_bug.cgi?id=100201) and [102318](https://bugs.freedesktop.org/show_bug.cgi?id=102318). This limits LLVM support to 4.0.1, so what I have is far different than Phoronix's environment.  But I gave it a shot anyway, I'll post here the binaries if build succeeds.

**Comment by pal1000** [comment]
Build successful but no SWR AVX512. I am either missing a parameter or Scons build lacks AVX512 support.
Anyway it's too early to ask for this so I am closing this issue.

**Comment by pal1000** [comment]
Mesa 17.3 is soon to be released. It checks 2 defines to determine wherever to enable AVX512 or not as seen in this commit: https://cgit.freedesktop.org/mesa/mesa/commit/?h=17.3&id=f8a572cdf0cf7fb52348adc7862a7ffc612180ef.

**Comment by oscarbg** [comment]
would be nice if you can get AVX512 builds for Mesa 17.3..
thanks..

**Comment by pal1000** [comment]
I'm afraid I have bad news regarding swr AVX512 build. I looked at Mesa3D source code inside the Sconscript used by swr in both [master](https://cgit.freedesktop.org/mesa/mesa/tree/src/gallium/drivers/swr/SConscript) and [17.3](https://cgit.freedesktop.org/mesa/mesa/tree/src/gallium/drivers/swr/SConscript?h=17.3) branches and there is no define or environment import for it. This looks like it simply isn't there.
Then I looked makefile.am and makefile.sources which are part of autools build and those have it. If you know how to build swr on linux you'd only have to pass HAVE_SWR_SKX and HAVE_SWR_KNL to configure.

**Comment by pal1000** [comment]
I found the discussion and bug report regarding cross compilation with mingw and as an irony the driver that fails to build is swr: https://bugs.freedesktop.org/show_bug.cgi?id=101614

**Comment by pal1000** [comment]
Sent an upstream feature request here:
https://bugs.freedesktop.org/show_bug.cgi?id=104166

**Comment by oscarbg** [comment]
Ji @pal1000 ,
just seen on mesa-dev "meson for SWR (on Linux) and Clover"
"https://lists.freedesktop.org/archives/mesa-dev/2017-December/179686.html"
which adds SWR support (albeit only Linux right now) to a new build generator (Meson) for Mesa which is getting mature fast..
seems SWR AVX512 is in but only for Linux also.. so hope Mesa gets Meson Windows SWR support soon (meson by itself supports generating Visual Studio projects so perhaps this Meson build files with little more work gets support)..
of course using Mingw will hit same bugs..
Finally don't know if may affect SWR AVX512 but seems Visual studio 2017 AVX512 was (is?) pretty buggy/not optimized right now.. see issue I opened here:
https://github.com/Mysticial/Flops/issues/16
author reported bugs and note there have been some Visual Studio 2017 point releases (from 15.3 at the time to 15.6 preview right now) so perhaps Microsoft have fixed major ones..

**Comment by pal1000** [comment]
After a lot of work and refactoring I managed to hook basic Meson support. Unfortunately my first attempt seam to indicate that MSVC is not yet supported. I get the following error during build config phase:

````
Build command: meson.exe . .\build\windows-x86 --backend=vs2017 -Dplatforms=windows

The Meson build system
Version: 0.45.1
Source dir: C:\Software\DEVELO~1\projects\mesa\mesa
Build dir: C:\Software\DEVELO~1\projects\mesa\mesa\build\windows-x86
Build type: native build
Program python found: YES (C:\Software\Development\projects\mesa\py3\python.EXE)
Project name: mesa
Native C compiler: cl (msvc 19.13.26131.1)
Native C++ compiler: cl (msvc 19.13.26131.1)
Build machine cpu family: x86
Build machine cpu: x86

meson.build:349:6: ERROR: Problem encountered: Cannot build GLX support without X11 platform support and at least one OpenGL API

A full log can be found at C:\Software\DEVELO~1\projects\mesa\mesa\build\windows-x86\meson-logs\meson-log.txt

Microsoft Windows [Version 10.0.16299.371]
(c) 2017 Microsoft Corporation. All rights reserved.

C:\Software\DEVELO~1\projects\mesa\mesa>

````

@oscarbg If you want to check why it fails and debug it, make sure Python 3,x is either in PATH or is installed with Python launcher, then you can apply the patch called `enablemeson` from `patches` folder (`git apply -v patches\enablemeson.patch`) and finally run the build script.  

The build script will attempt to build as you see above but then it opens an interactive command prompt to allow debugging.

The only build option I found relevant is `-DGLX_USE_WINDOWSGL`.  Unfortunately if I use it, it is not recognized and Meson throws a warning.

**Comment by oscarbg** [comment]
Thanks for effort.. are you using Mesa 18.0 or Mesa-dev in testing?

**Comment by pal1000** [comment]
I tried with Mesa-dev, though probably the same thing happens with 18.0.

**Comment by pal1000** [comment]
This issue mostly depends on #7 .

**Comment by pal1000** [comment]
I asked on mesa-dev mailing list about Meson support with MSVC and I got this response:
https://lists.freedesktop.org/archives/mesa-dev/2018-April/193436.html

And considering Dylan Baker is on to it this will take a while considering he is the release manager for Mesa 18.1.0. This will keep him away from Meson Windows support development on Mesa for 2 more weeks at least.

In a positive note I converted that enablemeson patch into a command line switch. It should be easier to attempt a Meson build this way.  The only thing left to do is to update the build script documentation as there are plenty of changes to make there. The Meson support is on hold until there is something upstream to work with. Right now there is nothing, not even on mesa master.

**Comment by oscarbg** [comment]
Thanks for all your effort..
can wait is not urgent.. 
I don't remember if asked before.. but will be easier to obtain SWR AVX512 Win builds if built via MingW GCC, Clang for Windows instead of MSVC?
just asking altough you aren't interested I may test other compilers if it's easier to build..

**Comment by pal1000** [comment]
> I don't remember if asked before.. but will be easier to obtain SWR AVX512 Win builds if built via MingW > GCC, Clang for Windows instead of MSVC?
> just asking altough you aren't interested I may test other compilers if it's easier to build..
Mingw GCC can't build SWR at all regardless of what OS the build machine is running due to a GCC stack alignment bug with AVX that mostly affects Windows binaries. You can Google for info on it. As for Clang, it would be indeed interesting to see if it is possible.
However on Windows, Clang runs an alternate toolset for MSVC.

**Comment by oscarbg** [comment]
thanks..
I see some MSVC AVX512 SWR work..
seems these patches are already merged:
https://cgit.freedesktop.org/mesa/mesa/commit/?id=fb20ae0374425ae3aff2a50a498c7e2b428632a4

**Comment by pal1000** [comment]
Well, I got to the point where I can build Mesa3D with Meson with LLVM linking and SWR driver included using [Dylan Baker's branch](https://gitlab.freedesktop.org/dbaker/mesa/tree/meson-windows), unfortunately enabling AVX512 support for SWR fails during build system configuration, it doesn't recognize SKX and KNL arguments which means no AVX512. Also even if you build only AVX and AVX2 you can't select SWR due to confusion in Meson build, basically `GALLIUM_DRIVER=swr` doesn't work for some reason.

**Comment by oscarbg** [comment]
@pal1000 finally!! that's good news for today..

**Comment by oscarbg** [comment]
are you interested on doing a release of Mesa 19.3-RC1?

**Comment by pal1000** [comment]
Yes. but will take a while as I need to rebuild LLVM with latest Visual Studio update.

**Comment by pal1000** [comment]
9.3.0-rc1 pre-release is out. Good luck.

**Comment by pal1000** [comment]
19.3..0 is built with experimental swrAVX512 support thanks to switch of MSVC binaries to Meson build and an out of tree patch. Official support is tracked [here](https://gitlab.freedesktop.org/mesa/mesa/issues/1990).
